client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -342,8 +343,8 @@ private void throwIfClosed() throws TxnFailedException {
         UUID txnId = txnSegments.getTxnId();
         Map<Segment, SegmentTransaction<Type>> transactions = new HashMap<>();
         for (Segment s : txnSegments.getSteamSegments().getSegments()) {
-            SegmentOutputStream out = outputStreamFactory.createOutputStreamForTransaction(s, txnId,
-                    config, txnSegments.getSteamSegments().getDelegationToken());
+            SegmentOutputStream out = outputStreamFactory.createOutputStreamForTransaction(s, txnId, config,
+                    DelegationTokenProviderFactory.create(txnSegments.getSteamSegments().getDelegationToken(), controller, s));
```

Comment:
Here we are creating many providers. This means when the token expires we will end up refreshing the token `numsegments` times rather than once.

Reply:
Fixed.